File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < n - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            int target = 0 - nums[i];
            int l = i + 1, r = n - 1;
            while (l < r) {
                if (nums[l] + nums[r] == target) {
                    res.add(new ArrayList<>(Arrays.asList(nums[i], nums[l], nums[r])));
                    while (l < r && nums[l] == nums[l - 1]) l++;
                    r--;
                    while (r > l && nums[r] == nums[r + 1]) r--;
                } else if (nums[l] + nums[r] < target) {
                    l++;
                } else {
                    r--;
                }                
            }
        }
        return res;
        /**      i        
             -4 -1 -1 0 1 2
                    l
                          r

            if (found)
                l++
                r--
            else if l + r < target
                l++
            else 
                r--

        */
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # -1 -1 0 1 2 4
        #       l
        #       r
        # 0 1 2
        nums.sort()
        n = len(nums)
        res = []
        for i in range(n - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            find = -nums[i]
            l, r = i + 1, n - 1
            while l < r:
                if nums[l] + nums[r] == find:
                    res.append([nums[i], nums[l], nums[r]])
                    
                if nums[l] + nums[r] > find:
                    r -= 1
                    while l < r and nums[r] == nums[r + 1]:
                        r -= 1
                else:
                    l += 1
                    while l < r and nums[l] == nums[l - 1]:
                        l += 1

        return res
